### PR TITLE
fix various TCP issues

### PR DESCRIPTION
**Pull request by soycode**
Depends on https://github.com/freedomjs/freedom/pull/267

### CODE REVIEW

**Comment by soycode** [comment]
Merged https://github.com/freedomjs/freedom/pull/267 as it was just small tweaks, and for this change to be clear - I don't want to merge it with the removal of onDisconnect. Rather I'd like to figure out why onDisconnect and pause/resume interact in a bad way (specifically it seems to cause the socket to never properly close and for the client socket to call onDisconnect repeatedly until JavaScript complains about recursive depth). So, more forthcoming.

**Comment by willscott** [comment]
this is ready to review?

**Comment by soycode** [comment]
I do believe it fixes the pause/resume test - there's still the issue where the `onDisconnect` dispatch causes uProxy to believe the client is disconnected, but that can be tackled in a separate pull request.

**Comment by soycode** [comment]
So it seems like the existing implementation has no `onclose` listener (https://developer.mozilla.org/en-US/docs/Web/API/WebSocket). Such a listener would receive a `CloseEvent` (https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent) that should let us delineate cases.

I'll take a hack at adding that, but am thinking of doing so in a separate pull request. Any objections to merging this request (which is mostly cleanup and passing through of continuations to ensure execution order)?

File: providers/client_socket.js [hunk 1]
```diff
@@ -148,7 +148,10 @@ ClientSocket.prototype.close = function() {
   // Delete transport so getInfo doesn't think we are connected
   delete this.transport;
   if (typeof this.onDisconnect === 'function') {
-    this.onDisconnect();
+    this.onDisconnect(continuation);
+    delete this.onDisconnect;
+  } else if (typeof continuation === 'function') {
+    continuation();
```
Comment (willscott) [inline providers/client_socket.js]: shouldn't continuation always be closed? not just if onDisconnect isn't defined?

File: providers/core.tcpsocket.js [hunk 2]
```diff
@@ -26,35 +26,44 @@ Socket_firefox.prototype.getInfo = function(continuation) {
...
 
+Socket_firefox.prototype.dispatchDisconnect = function(continuation, err) {
+  if (typeof err === 'undefined') {
+    err = {
+      'errcode': 'CONNECTION_CLOSED',
+      'message': 'Connection closed gracefully'
+    };
+  }
+  this.dispatchEvent('onDisconnect', err);
+  if (typeof continuation === 'function') {
```
Comment (willscott) [inline providers/core.tcpsocket.js]: this appears to never be called without a continuation, so you should be always able to call continuation(), and don't need to verify it's a function.

File: providers/client_socket.js (hunk 1 above)
Comment (soycode) [inline providers/client_socket.js]: If onDisconnect is defined the continuation gets passed through and should be called by it, at least as it's currently structured.

File: providers/core.tcpsocket.js (hunk 2 above)
Comment (soycode) [inline providers/core.tcpsocket.js]: Agreed, that's leftover debug code I believe, will change.

**Comment by soycode** [comment]
I'm making a new branch (soycode-tcprewrite) to try to tackle things a bit more properly (https://github.com/freedomjs/freedom-for-firefox/issues/64). I'll start it from the current state of this branch, though the first thing I'm doing is deleting the separate client/server files so it'll be pretty different regardless.

I'm still up for reviewing and tweaking this pull request to make it suitable for merging, especially if we can resolve https://github.com/freedomjs/freedom-for-firefox/issues/61 (so the next uProxy release can use updated freedom-for-firefox). In its current state this pull request I believe improves things (e.g. pause/resume passes), but doesn't fix the uProxy disconnect issue - thoughts/insight welcome.

**Comment by soycode** [comment]
Since I've basically just let this branch grow I've retitled the pull request appropriately. I believe this is ready for review, it should:

- Fix pause/resume TCP test by passing through continuations (and it turns out we do need that continuation existence check, or at least it helped prevent a crash at one point)
- Fix the onDisconnect dispatch stopping uProxy (https://github.com/freedomjs/freedom-for-firefox/issues/61)
- Some other small sundry spacing/style things

I'll work to test uProxy depending on freedom-for-firefox.jsm as built by this branch, and if it checks out I would encourage trying to get this merged before the next uProxy release (hopefully enabling metrics collection).

**Comment by willscott** [comment]
can you comment on test status after this pull request?

File: providers/server_socket.js [hunk 3]
```diff
@@ -50,9 +65,9 @@ ServerSocket.prototype.getInfo = function() {
   var port = nsIServerSocket.port;
   var localAddress = '127.0.0.1';
   var socketType = 'tcp';
-  var info = {socketType: socketType,
-              localAddress: localAddress,
-              localPort: port};
+  var info = { socketType   : socketType,
```
Comment (willscott) [inline providers/server_socket.js]: i don't think we have this spacing convention anywhere else.

Comment (soycode) [inline providers/server_socket.js]: Yeah I didn't like it much either, just tried to do a minimal change from what it was. Have made it more normal now.

**Comment by soycode** [comment]
Did more testing and decided for the time being to abandon this pull request - it's just a bit too kludgy. What we really need immediately is: https://github.com/freedomjs/freedom-for-firefox/pull/65

I'll probably revisit some ideas from here later, but for now want to do the smallest/simplest change possible to get uProxy updated.

**Pull request closed without merging**